src/neptune/internal/operation_processors/lazy_operation_processor_wrapper.py

```diff
@@ -0,0 +1,112 @@
+#
+# Copyright (c) 2024, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+
+__all__ = ("LazyOperationProcessorWrapper",)
+
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    Optional,
+    TypeVar,
+)
+
+from neptune.core.components.abstract import Resource
+from neptune.core.components.operation_storage import OperationStorage
+from neptune.internal.operation import Operation
+from neptune.internal.operation_processors.operation_processor import OperationProcessor
+
+RT = TypeVar("RT")
+
+
+def trigger_evaluation(method: Callable[..., RT]) -> Callable[..., RT]:
+    def _wrapper(self: LazyOperationProcessorWrapper, *args: Any, **kwargs: Any) -> RT:
+        if self._operation_processor is None:
+            self._operation_processor = self._operation_processor_getter()
+            if self._post_trigger_side_effect is not None:
+                self._post_trigger_side_effect()
+        return method(self, *args, **kwargs)
+
+    return _wrapper
+
+
+def noop_if_not_triggered(method: Callable[..., RT]) -> Callable[..., RT]:
+    def _wrapper(self: LazyOperationProcessorWrapper, *args: Any, **kwargs: Any) -> RT:
+        if self._operation_processor is not None:
+            return method(self, *args, **kwargs)
+
+    return _wrapper
+
+
+class LazyOperationProcessorWrapper(OperationProcessor):
+    def __init__(
+        self,
+        operation_processor_getter: Callable[[], OperationProcessor],
+        post_trigger_side_effect: Optional[Callable[[], Any]] = None,
+    ):
+        self._operation_processor_getter = operation_processor_getter
+        self._post_trigger_side_effect = post_trigger_side_effect
+        self._operation_processor: OperationProcessor = None  # type: ignore
+
+    def evaluated(self) -> bool:
```

Comment:
Take a look at the usage of `evaluated` in my PR: https://github.com/neptune-ai/neptune-client/pull/1628/files#diff-aee064cd8dc933c9e3a255fd43e594782bf4146bb594326cbb8e3c66a7debc62R500 . If I understand correctly, here we will always have a `stop`-related messages that are probably not needed. Especially, we rather don't want to log anything to stdout if OperationProcessor wasn't initialized.

Reply:
+ How about a `@property`?

Reply:
I will investigate excessive stop messages as I was not aware. Maybe there is another way to get rid of them.

I don't like using `evaluated` in `metadata_container` as its outside of OperationProcessor interface and it introduces type errors if `self._op_processor` is not defined as `LazyOperationProcessorWrapper`. 

Thats another reason why I made decision discussed [here](https://github.com/neptune-ai/neptune-client/pull/1649#discussion_r1490636799)

Reply:
Yeah, for sure we should do something more feasible with `evaluated` than I did in Proof-of-Concept :wink:. Just be aware that we don't want to have any messages if the operation processor wasn't created (this brings us mostly to stop messages).

Reply:
UPDATE: indeed fork spawned childs are cousin additional logs.

If the intended behaviour is not logging shutdown messages if `LazyOperationProcessor` was not triggered, maybe we can add `shutdown hook` that will log this messages before or after `operationProcessor.stop()` function.

We could also for each `metadata_container` process keep var to know if we are parent (responsible for logging shutdown messages) or child which should be silent.

Let me know what you think about that :D

Reply:
> If the intended behaviour is not logging shutdown messages if LazyOperationProcessor was not triggered, maybe we can add shutdown hook that will log this messages before or after operationProcessor.stop() function.

So you may assume that there is no need to print anything except the logs from `OperationProcessor.stop()` itself from a forked Run. So:
- No need to print info about background jobs shuting down
- No need to print info about URL to metadata

Does it help you?

Reply:
This means that you don't care about evaluated/non-evaluated as it's a part of OperationProcessor implementation and rather just store info about being forked/or not that could be a part of `MetadataContainer`.

Reply:
yep, ty :)